test/jdk/jdk/crac/fileDescriptors/PatchModuleTest.java

```diff
@@ -0,0 +1,140 @@
+        Path classesDir = Files.createTempDirectory("patch-classes");
+        boolean compiled = CompilerUtils.compile(
+                srcDir,
+                classesDir,
+                "--patch-module=java.base=" + srcDir.toString());
+        assertTrue(compiled, "Compilation of patch class with --patch-module failed");
+
+        // 3. Create the patch JAR file from the compiled class
+        File patchJar = File.createTempFile("patch", ".jar");
+        patchJar.deleteOnExit();
+
+        String[] patchJarArgs = {
+                "--create",
+                "--file", patchJar.getAbsolutePath(),
+                "-C", classesDir.toString(), "."
+        };
+        sun.tools.jar.Main jarTool = new sun.tools.jar.Main(System.out, System.err, "jar");
+        if (!jarTool.run(patchJarArgs)) {
+            throw new Exception("Failed to create patch JAR: " + Arrays.toString(patchJarArgs));
+        }
+
+        // 4. Build the process with --patch-module
+        CracBuilder builder = new CracBuilder()
+                .captureOutput(true)
+                .printResources(true)
+                .vmOption("--patch-module=java.base=" + patchJar.getAbsolutePath());
+
+        // 5. Start the checkpoint process
+        CracProcess checkpointProcess = builder.startCheckpoint();
+
+        // 6. Analyze the output of the checkpointing process to verify fixes
+        OutputAnalyzer checkpointOutput = checkpointProcess.outputAnalyzer();
+
+        // Verify that the checkpoint process did not fail
+        checkpointOutput.shouldNotContain("CheckpointException");
+
+        // Verify Java-level fix: PersistentJarFile is used
+        checkpointOutput.shouldContain(patchJar.getAbsolutePath() + " is recorded as always available on restore");
+
+        // Verify Native-level fix: check_fds reports the FD as ok
+        checkpointOutput.shouldMatch("JVM: FD fd=[0-9]+ type=regular path=\"" + patchJar.getAbsolutePath()
+                + "\" OK: claimed by patch-module");
+
+        // Let the checkpoint finish
+        checkpointProcess.waitForCheckpointed();
+
+        // 7. Perform restore
+        builder.doRestore();
+    }
+
+    @Override
+    public void exec() throws Exception {
+        // Verify the patch is active by loading and calling the new class
+        String infoBefore = invokePatchedMethod();
+        assertEquals(PATCH_MESSAGE, infoBefore, "Patch should be active before checkpoint.");
+
+        // Perform checkpoint and restore
+        Core.checkpointRestore();
+
+        // Verify the patch is still active after restore
+        String infoAfter = invokePatchedMethod();
+        assertEquals(PATCH_MESSAGE, infoAfter, "Patch should remain active after restore.");
+    }
+
+    private String invokePatchedMethod() throws Exception {
+        Class<?> patcherClass = Class.forName(PATCHED_CLASS_FULL_NAME);
+        return (String) patcherClass.getMethod("getInfo").invoke(null);
+    }
+}
```

Comment:
I haven't looked through the test but if you will be applying Radim's suggestions please also add a newline at the end of file

Reply:
Good catch, thanks! I've pushed a commit with the added newline.

I've also left a comment on Radim's suggestion explaining why I'm sticking with the criu engine for this test to ensure a full physical restore is validated.

Appreciate the review!